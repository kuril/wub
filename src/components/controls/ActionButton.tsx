import * as React from "react";
import { connect } from 'react-redux';
import { RootState } from "types/wub";
import { Dispatch } from "redux";
import { createArticleStartLoadingAction, createArticleSetResultAction } from "../../ducks";
import classnames = require("classnames");
import { loadArticles } from "../../api";

type ButtonStateProps = { loading: boolean }
type ButtonDispatchProps = { loadArticles: () => void }
type ButtonProps = ButtonStateProps & ButtonDispatchProps;

const ActionButton: React.SFC<ButtonProps> = ({ loading, loadArticles }) => {
    return (
        <button className={classnames(
            "button", "is-primary",
            { ["is-loading"]: loading })}
        onClick={loadArticles} disabled={loading} >
            Load Articles
    </button>
    )
}

const mapStateToProps = (state: RootState) => ({
    loading: state.article.isLoading,
})

const mapDispatchToProps = (dispatch: Dispatch) => ({
    loadArticles: () => {
        dispatch(createArticleStartLoadingAction())
        loadArticles().then(result => dispatch(createArticleSetResultAction(result)))
    }
})

export const ActionButtonConnected =
    connect<ButtonStateProps, ButtonDispatchProps, {}>(
        mapStateToProps, mapDispatchToProps
    )(ActionButton)