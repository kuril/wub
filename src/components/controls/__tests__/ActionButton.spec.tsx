import * as React from "react";
import * as TestRenderer from "react-test-renderer";
import { ActionButton } from "../ActionButton";

const loadArticles = jest.fn()

const renderWithProps = (loading: boolean) => TestRenderer.create(
    <ActionButton loading={loading} loadArticles={loadArticles} />).toJSON()

describe("ActionButton component", () => {
    it("renders correctly", () => {
        expect(renderWithProps(false)).toMatchSnapshot();
    })
    it("renders correctly when loading is true", () => {
        expect(renderWithProps(true)).toMatchSnapshot();
    })
    it("loads articles", () => {
        renderWithProps(false).props.onClick()
        expect(loadArticles.mock.calls.length).toBe(1)
    })
})