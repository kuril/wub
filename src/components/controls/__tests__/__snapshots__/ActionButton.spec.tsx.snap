// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionButton component renders correctly 1`] = `
<button
  className="button is-primary"
  disabled={false}
  onClick={[MockFunction]}
>
  Load Articles
</button>
`;

exports[`ActionButton component renders correctly when loading is true 1`] = `
<button
  className="button is-primary is-loading"
  disabled={true}
  onClick={[MockFunction]}
>
  Load Articles
</button>
`;
